### ISSUE TITLE
Bug when translating functions that contain an "if" (in my case an implicit if)

### ISSUE THREAD

**Issue by rtoy** [rtoy/maxima#2660]
Imported from SourceForge on 2024-07-06 17:12:09
Created by **[peterpall](https://sourceforge.net/u/peterpall/)** on 2018-03-18 07:41:36
Original: https://sourceforge.net/p/maxima/bugs/3412

---

In the following example uncommenting the compile step drastically changes the result.

~~~
(%i40)	f(x):=if cabs(1/(x+1)) < 1 then 1/(x+1) else 1;
(%o40)	f(x):=if \#\{Lisp function\}(1/(x+1))<1 then 1/(x+1) else 1
	/* compile(f);*/
(%i41)	draw3d(contour='map,
	    proportional_axes=xy,
	    nticks=100,
	    contour_levels=20,
	    explicit(f(x+%i*y),x,-2,2,y,-2,2)
)$
~~~
Comments from Robert Dodier:
Actually since f is called just once (when gathering the arguments for draw3d) it's easy to see the problem. Non-translated: 
~~~
f(x+%i*y) => if
... then .... 
~~~
Translated: 
~~~
f(x+%i*y) => 1/(1 + stuff)
~~~
i.e. no "if".

Looking at the translated code shows the problem. 

~~~
(%i6) translate(f);
(%i7) :lisp (symbol-function '$f)
#<FUNCTION $F ($X) (DECLARE (IN-DEFUN $F) (SPECIAL $X))
  (BLOCK $F (COND ((IS-BOOLE-CHECK (MLSP (SIMPLIFY ($CABS (DIV 1 (ADD*
  $X 1)))) 1)) (DIV 1 (ADD* $X 1))) (T 1)))>
~~~

The COND only has clauses for the two branches of the "if". There should be a third clause, to return ((MCOND) ...) (i.e. an "if" expression) when IS-BOOLE-CHECK returns $UNKNOWN instead of T or NIL.

The current translation is OK when prederror = true since then, IS-BOOLE-CHECK returns only T or NIL (I think). But when prederror = false (default), it's not correct.

Looking at the code (see (DEF%TR MCOND ...) in src/transl.lisp), it seems easy enough to fix.

Comment from Kris Katterjohn: You can see the function body in SBCL with
~~~
:lisp (describe #'$f).
~~~

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-06 17:12:11
Created by **[macrakis](https://sourceforge.net/u/macrakis/)** on 2018-03-23 18:42:18
Original: https://sourceforge.net/p/maxima/bugs/3412/#2578

---

Getting conditionals with prederror:false to translate correctly is messy and difficult.

Yes, of course you can get (if x>0 then 1) to translate to (temp: maybe(x>0), if temp='unknown then '(if x>0 then 1) elseif temp then 1) but that doesn't help with if clauses that have side effects:

      ( if x<0 then y:3 , ... )
        
 How about
 
        for i in k while i>0 do ...
        
 Should that translate to
 
       block([i,ktemp:k,reltemp],
           while (i: first(ktemp), ktemp:rest(ktemp), (reltemp:maybe(i>0))=true)
                 do ...,
           if reltemp=false then 'done else '(for i in ktemp while i>0 do ...))
 
   ? It could, but what's the point?
   
I think the user is less likely to blunder into weird cases if we simply give an error if a conditional returns unknown in translated code, regardless of the value of prederrror.

The OP simply needed to write '(f(x+%i*y)) instead of f(x+%i*y).

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-06 17:12:15
Created by **[peterpall](https://sourceforge.net/u/peterpall/)** on 2018-04-01 17:46:24
Original: https://sourceforge.net/p/maxima/bugs/3412/#2578/901d

---

It definitively is messy. But currently the compiler doesn't support "if" at all, in a way: With prederror:false it omits obviously wrong code and with prederror:true it aborts with an error if  the "if" is actually needed.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-06 17:12:18
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2019-02-24 02:27:36
Original: https://sourceforge.net/p/maxima/bugs/3412/#2578/33bb

---

Since this topic has come up again, I thought about it again, and at this point I'm inclined to agree that translated code should give an error when evaluating the conditional expression in if-then returns `unknown`.

Maybe translated if-then expressions can be wrapped in `(let (($prederror t)) ...)`.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-06 17:12:22
Created by **[peterpall](https://sourceforge.net/u/peterpall/)** on 2019-02-24 08:48:35
Original: https://sourceforge.net/p/maxima/bugs/3412/#2578/33bb/e8b5

---

If a translated function raised an error if an "if" encounters an "unknown" at run-time I wouldn't be against that.

But setting `prederror` to true wouldn't resolve the original problem but only inform the result of the function would have been invalid, anyway, which is but one step forward:
~~~
(%i1)	prederror:true;
(%o1)	true

(%i2)	sgn(x):=block([],if is(x>0.0) then return(1) elseif is(x<0.0) then return(-1) else return(0.0)); 
(%o2)	sgn(x):=block([],if is(x>0.0) then return(1) elseif 
is(x<0.0) then return(-1) else return(0.0))

(%i3)	translate(sgn);
(%o3)	[sgn]

(%i6)	sgn(1);sgn(0); sgn(-2.1); 
(%o4)	1
(%o5)	0.0
(%o6)	-1

(%i7)	compile(sgn);
(%o7)	[sgn]
(%i8)	sgn(0);
(%o8)	0.0
(%i9)	sgn(-1);
(%o9)	-1
(%i11)	sgn(1);
(%o11)	1

(%i12)	wxdraw2d(explicit(sgn(x),x,-1,1));
Unable to evaluate predicate x>0

 -- an error. To debug this try: debugmode(true);
~~~

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-06 17:12:25
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2019-02-24 23:03:43
Original: https://sourceforge.net/p/maxima/bugs/3412/#fcc1

---

- **labels**:  --> translate, compile, if
- **status**: open --> closed

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-06 17:12:29
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2019-02-24 23:03:43
Original: https://sourceforge.net/p/maxima/bugs/3412/#602c

---

I've changed the translator for "if" (i.e. DEF%TR MCOND in src/transl.lisp) to emit (LET (($PREDERROR T)) ...) wrapped around the translated "if". Therefore translated "if" expressions have  the behavior suggested by Stavros. I've pushed commit 74c138c with that change. 

I believe this approach is an improvement, since it prevents incorrect behavior, but it is still a little unsatisfying to me. In the event that we devise a way to ensure the same behavior in interpreted and translated "if" expressions, we can revert this change, til then we should let it stand, I believe.
